python/dgl/transforms/module.py

```diff
@@ -98,6 +100,162 @@ def __call__(self, g):
+        Default: `None`.
+    edge_feat_names : list of str, optional
+        The names of the edge features to be normalized. If None, all edge features will be normalized.
+        Default: `None`.
+
+    Example
+    -------
+
+    >>> import dgl
+    >>> import torch
+    >>> from dgl import NormalizeFeatures
+
+    Case1: Normalize features of a homogenous graph.
+
+    >>> transform = NormalizeFeatures()
+    >>> g = dgl.rand_graph(5, 20)
+    >>> g.ndata['h'] = torch.randn((g.num_nodes(), 5))
+    >>> print(g.ndata['h'].sum(1))
+    tensor([-3.0586,  3.4974,  3.1509,  1.5805,  0.4890])
+    >>> g.edata['w'] = torch.randn((g.num_edges, 5))
+    >>> print(g.edata['w'].sum(1))
+    tensor([ 2.9284, -3.8341,  1.5087,  0.8673, -2.5115,  2.4751,  0.1427,  1.7180,
+            2.7705,  1.0600, -0.7126,  0.1072, -0.8159,  1.2082,  2.0327, -2.3323,
+            -1.2495,  1.9458, -1.4240, -1.3575])
+    >>> new_g = transform(g)
+    >>> print(new_g.ndata['h'].sum(1))
+    tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000])
+    >>> print(new_g.edata['w'].sum(1))
+    tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
+            1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
+            1.0000, 1.0000])
+
+    Case2: Normalize features of a heterogeneous graph.
+
+    >>> transform = NormalizeFeatures()
+    >>> g = dgl.heterograph({
+    ...     ('user', 'follows', 'user'): (torch.tensor([1, 2]), torch.tensor([3, 4])),
+    ...     ('player', 'plays', 'game'): (torch.tensor([2, 2]), torch.tensor([1, 1]))
+    ... })
+    >>> g.ndata['h'] = {'game': torch.randn(2, 5), 'player': torch.randn(3, 5)}
+    >>> print(g.ndata['h']['game'].sum(1), g.ndata['h']['player'].sum(1))
+    tensor([ 4.4201, -4.0683]) tensor([-2.2460, -1.1204, -2.0254])
+    >>> g.edata['w'] = {('user', 'follows', 'user'): torch.randn(2, 5), ('player', 'plays', 'game'): torch.randn(2, 5)}
+    >>> print(g.edata['w'][('user', 'follows', 'user')].sum(1), g.edata['w'][('player', 'plays', 'game')].sum(1))
+    tensor([-1.2663,  3.3789]) tensor([4.1371, 1.4743])
+    >>> new_g = transform(g)
+    >>> print(new_g.ndata['h']['game'].sum(1), new_g.ndata['h']['player'].sum(1))
+    tensor([1.0000, 1.0000]) tensor([1.0000, 1.0000, 1.0000])
+    >>> print(new_g.edata['w'][('user', 'follows', 'user')].sum(1), new_g.edata['w'][('player', 'plays', 'game')].sum(1))
+    tensor([1.0000, 1.0000]) tensor([1.0000, 1.0000])
+    """
+    def __init__(self, node_feat_names=None, edge_feat_names=None):
+        self.node_feat_names = node_feat_names
+        self.edge_feat_names = edge_feat_names
+
+    def __call__(self, g):
+        def normalize(feat):
+            feat = feat - feat.min()
+            feat.div_(feat.sum(dim=-1, keepdim=True).clamp_(min=1.))
+            return feat
+
+        if self.node_feat_names is None:
+            self.node_feat_names = g.ndata.keys()
```

Comment:
Does it work if nodes/edges of different types have different feature names?

Reply:
Yes, it does work. For heterogeneous graph, the `g.ndata` stores node data like this `{'feat_name': {'node_type': tensor}}`. So `g.ndata.keys()` will get all the feature names.